File: lib/services/bidirectional_sync_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io' show Platform;
import 'package:flutter/foundation.dart' show defaultTargetPlatform, TargetPlatform, debugPrint, ChangeNotifier;
import 'package:device_info_plus/device_info_plus.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'api_service.dart';
import 'local_database_service.dart';
import 'connectivity_service.dart';
import '../models/asistencia_model.dart';
import '../models/presencia_model.dart';

/// Servicio de sincronización bidireccional con versionado y manejo de conflictos
class BidirectionalSyncService {
  static final BidirectionalSyncService _instance = BidirectionalSyncService._internal();
  factory BidirectionalSyncService() => _instance;
  BidirectionalSyncService._internal();

  final ApiService _apiService = ApiService();
  final LocalDatabaseService _localDb = LocalDatabaseService();
  final ConnectivityService _connectivityService = ConnectivityService();

  String? _deviceId;
  String? _syncToken;
  DateTime? _lastSyncTime;
  bool _isSyncing = false;
  int _conflictCount = 0;

  // Getters
  String? get deviceId => _deviceId;
  String? get syncToken => _syncToken;
  DateTime? get lastSyncTime => _lastSyncTime;
  bool get isSyncing => _isSyncing;
  int get conflictCount => _conflictCount;

  /// Inicializar servicio de sincronización
  Future<void> initialize() async {
    await _connectivityService.initialize();
    await _initializeDeviceId();
    await _registerDevice();
    await _loadSyncState();
  }

  /// Inicializar ID del dispositivo
  Future<void> _initializeDeviceId() async {
    final prefs = await SharedPreferences.getInstance();
    _deviceId = prefs.getString('device_id');

    if (_deviceId == null) {
      final deviceInfo = DeviceInfoPlugin();
      String deviceIdentifier;
      
      if (Platform.isAndroid) {
        final androidInfo = await deviceInfo.androidInfo;
        deviceIdentifier = androidInfo.id;
      } else if (Platform.isIOS) {
        final iosInfo = await deviceInfo.iosInfo;
        deviceIdentifier = iosInfo.identifierForVendor ?? 'ios-unknown';
      } else {
        deviceIdentifier = 'unknown-${DateTime.now().millisecondsSinceEpoch}';
      }

      _deviceId = deviceIdentifier;
      await prefs.setString('device_id', _deviceId!);
    }
  }

  /// Registrar dispositivo en el servidor
  Future<void> _registerDevice() async {
    try {
      final deviceInfo = DeviceInfoPlugin();
      String deviceName = 'Unknown';
      String deviceType = 'mobile';
      String appVersion = '1.0.0';
      
      try {
        final packageInfo = await PackageInfo.fromPlatform();
        appVersion = packageInfo.version;
      } catch (e) {
        debugPrint('Error obteniendo versión de app: $e');
      }
      
    if (defaultTargetPlatform == TargetPlatform.android) {
      final androidInfo = await deviceInfo.androidInfo;
      deviceName = '${androidInfo.manufacturer} ${androidInfo.model}';
      deviceType = 'mobile';
    } else if (defaultTargetPlatform == TargetPlatform.iOS) {
      final iosInfo = await deviceInfo.iosInfo;
      deviceName = '${iosInfo.name} ${iosInfo.model}';
      deviceType = 'mobile';
    }

      final response = await _apiService.post(
        '/sync/register-device',
        {
          'device_id': _deviceId,
          'device_name': deviceName,
          'device_type': deviceType,
          'app_version': appVersion,
        },
      );

      if (response['success'] == true) {
        _syncToken = response['sync_token'];
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString('sync_token', _syncToken!);
      }
    } catch (e) {
      debugPrint('Error registrando dispositivo: $e');
    }
  }

  /// Cargar estado de sincronización
  Future<void> _loadSyncState() async {
    final prefs = await SharedPreferences.getInstance();
    _syncToken = prefs.getString('sync_token');
    final lastSyncString = prefs.getString('last_sync_time');
    if (lastSyncString != null) {
      _lastSyncTime = DateTime.parse(lastSyncString);
    }
    _conflictCount = prefs.getInt('conflict_count') ?? 0;
  }

  /// Guardar estado de sincronización
  Future<void> _saveSyncState() async {
    final prefs = await SharedPreferences.getInstance();
    if (_syncToken != null) {
      await prefs.setString('sync_token', _syncToken!);
    }
    if (_lastSyncTime != null) {
      await prefs.setString('last_sync_time', _lastSyncTime!.toIso8601String());
    }
    await prefs.setInt('conflict_count', _conflictCount);
  }

  /// Realizar sincronización bidireccional completa
  Future<SyncResult> performBidirectionalSync({bool forceSync = false}) async {
    if (_isSyncing && !forceSync) {
      return SyncResult(
        success: false,
        error: 'Sincronización ya en curso',
      );
    }

    if (!_connectivityService.isOnline && !forceSync) {
      return SyncResult(
        success: false,
        error: 'Sin conexión a internet',
      );
    }

    _isSyncing = true;
    notifyListeners();

    try {
      // 1. Obtener cambios pendientes del cliente
      final clientChanges = await _getPendingClientChanges();

      // 2. Realizar sincronización bidireccional
      final response = await _apiService.post(
        '/sync/bidirectional',
        {
          'device_id': _deviceId,
      'device_info': await _getDeviceInfo(),
          'last_sync': _lastSyncTime?.toIso8601String(),
          'client_changes': clientChanges,
        },
      );

      if (response['success'] != true) {
        throw Exception('Error en sincronización: ${response['error']}');
      }

      // 3. Procesar cambios del servidor (pull)
      final serverChanges = response['server_changes'];
      await _applyServerChanges(serverChanges);

      // 4. Procesar resultados de upload (push)
      final uploadResults = response['upload_results'];
      await _processUploadResults(uploadResults);

      // 5. Actualizar estado
      _lastSyncTime = DateTime.now();
      _conflictCount = uploadResults['conflicts']?.length ?? 0;
      await _saveSyncState();

      _isSyncing = false;

      return SyncResult(
        success: true,
        syncedCount: uploadResults['synced']?.length ?? 0,
        conflictCount: _conflictCount,
        serverChangesCount: serverChanges['changes']?.length ?? 0,
      );
    } catch (e) {
      _isSyncing = false;
      return SyncResult(
        success: false,
        error: e.toString(),
      );
    }
  }

  /// Obtener cambios pendientes del cliente
  Future<List<Map<String, dynamic>>> _getPendingClientChanges() async {
    final changes = <Map<String, dynamic>>[];

    // Obtener asistencias pendientes
    final pendingAsistencias = await _localDb.getPendingAsistencias();
    for (final asistencia in pendingAsistencias) {
      final version = await _localDb.getRecordVersion('asistencias', asistencia.id);
      changes.add({
        'collection': 'asistencias',
        'record_id': asistencia.id,
        'operation': 'create',
        'data': asistencia.toJson(),
        'version': version,
        'hash': _calculateHash(asistencia.toJson()),
      });
    }

    // Obtener presencia pendiente
    final pendingPresencia = await _localDb.getPendingPresencia();
    for (final presencia in pendingPresencia) {
      final version = await _localDb.getRecordVersion('presencia', presencia.id);
      changes.add({
        'collection': 'presencia',
        'record_id': presencia.id,
        'operation': 'create',
        'data': presencia.toJson(),
        'version': version,
        'hash': _calculateHash(presencia.toJson()),
      });
    }

    return changes;
  }

  /// Aplicar cambios del servidor
  Future<void> _applyServerChanges(Map<String, dynamic> serverChanges) async {
    final changes = serverChanges['changes'] as List?;
    if (changes == null) return;

    for (final change in changes) {
      try {
        await _applyServerChange(change);
      } catch (e) {
        debugPrint('Error aplicando cambio del servidor: $e');
      }
    }
  }

  /// Aplicar un cambio individual del servidor
  Future<void> _applyServerChange(Map<String, dynamic> change) async {
    final collection = change['collection'] as String;
    final recordId = change['record_id'] as String;
    final operation = change['operation'] as String;
    final data = change['data'] as Map<String, dynamic>;
    final version = change['version'] as int;

    switch (collection) {
      case 'asistencias':
        if (operation == 'update' || operation == 'create') {
          final asistencia = AsistenciaModel.fromJson(data);
          await _localDb.saveAsistencia(asistencia, syncStatus: 'synced');
          await _localDb.setRecordVersion('asistencias', recordId, version);
        } else if (operation == 'delete') {
          await _localDb.deleteAsistencia(recordId);
        }
        break;
      case 'presencia':
        if (operation == 'update' || operation == 'create') {
          final presencia = PresenciaModel.fromJson(data);
          await _localDb.savePresencia(presencia, syncStatus: 'synced');
          await _localDb.setRecordVersion('presencia', recordId, version);
        } else if (operation == 'delete') {
          await _localDb.deletePresencia(recordId);
        }
        break;
    }
  }

  /// Procesar resultados de upload
  Future<void> _processUploadResults(Map<String, dynamic> uploadResults) async {
    // Procesar cambios sincronizados
    final synced = uploadResults['synced'] as List?;
    if (synced != null) {
      for (final item in synced) {
        final collection = item['collection'] as String;
        final recordId = item['record_id'] as String;
        final version = item['version'] as int;

        await _localDb.markAsSynced(collection, recordId);
        await _localDb.setRecordVersion(collection, recordId, version);
      }
    }

    // Procesar conflictos
    final conflicts = uploadResults['conflicts'] as List?;
    if (conflicts != null && conflicts.isNotEmpty) {
      _conflictCount = conflicts.length;
      await _handleConflicts(conflicts);
    }
  }

  /// Manejar conflictos
  Future<void> _handleConflicts(List<dynamic> conflicts) async {
    for (final conflict in conflicts) {
      final conflictId = conflict['conflict_id'] as String;
      final collection = conflict['collection'] as String;
      final recordId = conflict['record_id'] as String;
      final serverData = conflict['server_data'] as Map<String, dynamic>?;
      final clientVersion = conflict['client_version'] as int;
      final serverVersion = conflict['server_version'] as int;

      // Guardar conflicto para resolución
      await _localDb.saveConflict(
        conflictId: conflictId,
        collection: collection,
        recordId: recordId,
        serverData: serverData,
        clientVersion: clientVersion,
        serverVersion: serverVersion,
      );

      // Aplicar estrategia de resolución por defecto (last_write_wins)
      // El usuario puede cambiar esto después
      await resolveConflict(
        conflictId: conflictId,
        strategy: ConflictResolutionStrategy.lastWriteWins,
      );
    }
  }

  /// Resolver conflicto
  Future<bool> resolveConflict({
    required String conflictId,
    required ConflictResolutionStrategy strategy,
    Map<String, dynamic>? resolutionData,
  }) async {
    try {
      final response = await _apiService.post(
        '/sync/conflicts/$conflictId/resolve',
        {
          'strategy': strategy.toString().split('.').last,
          'resolved_by': _deviceId,
          'resolution_data': resolutionData,
        },
      );

      if (response['success'] == true) {
        await _localDb.markConflictResolved(conflictId);
        _conflictCount = (_conflictCount > 0) ? _conflictCount - 1 : 0;
        await _saveSyncState();
        return true;
      }

      return false;
    } catch (e) {
      debugPrint('Error resolviendo conflicto: $e');
      return false;
    }
  }

  /// Obtener conflictos pendientes
  Future<List<ConflictInfo>> getPendingConflicts() async {
    try {
      final response = await _apiService.get(
        '/sync/conflicts?device_id=$_deviceId',
      );

      if (response['success'] == true) {
        final conflicts = response['conflicts'] as List;
        return conflicts.map((c) => ConflictInfo.fromJson(c)).toList();
      }

      return [];
    } catch (e) {
      debugPrint('Error obteniendo conflictos: $e');
      return [];
    }
  }

  /// Obtener información del dispositivo
  Future<Map<String, dynamic>> _getDeviceInfo() async {
    final deviceInfo = DeviceInfoPlugin();
    String deviceName = 'Unknown';
    String deviceType = 'mobile';
    String appVersion = '1.0.0';
    
    try {
      final packageInfo = await PackageInfo.fromPlatform();
      appVersion = packageInfo.version;
    } catch (e) {
      debugPrint('Error obteniendo versión de app: $e');
    }
    
    if (defaultTargetPlatform == TargetPlatform.android) {
      final androidInfo = await deviceInfo.androidInfo;
      deviceName = '${androidInfo.manufacturer} ${androidInfo.model}';
      deviceType = 'mobile';
    } else if (defaultTargetPlatform == TargetPlatform.iOS) {
      final iosInfo = await deviceInfo.iosInfo;
      deviceName = '${iosInfo.name} ${iosInfo.model}';
      deviceType = 'mobile';
    }

    return {
      'device_name': deviceName,
      'device_type': deviceType,
      'app_version': appVersion,
    };
  }

  /// Calcular hash de datos
  String _calculateHash(Map<String, dynamic> data) {
    final jsonString = jsonEncode(data);
    // Hash simple usando código de caracteres
    int hash = 0;
    for (int i = 0; i < jsonString.length; i++) {
      hash = ((hash << 5) - hash) + jsonString.codeUnitAt(i);
      hash = hash & hash; // Convert to 32bit integer
    }
    return hash.toString();
  }
}

/// Resultado de sincronización
class SyncResult {
  final bool success;
  final String? error;
  final int syncedCount;
  final int conflictCount;
  final int serverChangesCount;

  SyncResult({
    required this.success,
    this.error,
    this.syncedCount = 0,
    this.conflictCount = 0,
    this.serverChangesCount = 0,
  });
}

/// Estrategias de resolución de conflictos
enum ConflictResolutionStrategy {
  serverWins,
  clientWins,
  merge,
  lastWriteWins,
  manual,
}

/// Información de conflicto
class ConflictInfo {
  final String conflictId;
  final String collection;
  final String recordId;
  final int clientVersion;
  final int serverVersion;
  final Map<String, dynamic>? serverData;
  final Map<String, dynamic>? clientData;

  ConflictInfo({
    required this.conflictId,
    required this.collection,
    required this.recordId,
    required this.clientVersion,
    required this.serverVersion,
    this.serverData,
    this.clientData,
  });

  factory ConflictInfo.fromJson(Map<String, dynamic> json) {
    return ConflictInfo(
      conflictId: json['_id'] ?? json['conflict_id'],
      collection: json['collection_name'] ?? json['collection'],
      recordId: json['record_id'],
      clientVersion: json['client_version'] ?? json['version'],
      serverVersion: json['server_version'],
      serverData: json['server_data'],
      clientData: json['data'],
    );
  }
}

File: backend/services/bidirectional_sync_service.js
/**
 * Servicio de Sincronización Bidireccional
 * Maneja sincronización bidireccional, versionado y resolución de conflictos
 */

const crypto = require('crypto');

class BidirectionalSyncService {
  constructor(DataVersionModel, DeviceSyncModel, PendingChangeModel, AsistenciaModel, PresenciaModel) {
    this.DataVersion = DataVersionModel;
    this.DeviceSync = DeviceSyncModel;
    this.PendingChange = PendingChangeModel;
    this.Asistencia = AsistenciaModel;
    this.Presencia = PresenciaModel;
  }

  /**
   * Calcula hash de un objeto para detección de cambios
   */
  calculateHash(data) {
    const str = JSON.stringify(data);
    return crypto.createHash('md5').update(str).digest('hex');
  }

  /**
   * Obtiene o crea versión de un registro
   */
  async getOrCreateVersion(collectionName, recordId, data = null) {
    try {
      let version = await this.DataVersion.findOne({
        collection_name: collectionName,
        record_id: recordId
      });

      if (!version) {
        const { v4: uuidv4 } = require('uuid');
        const hash = data ? this.calculateHash(data) : '';
        
        version = new this.DataVersion({
          _id: uuidv4(),
          collection_name: collectionName,
          record_id: recordId,
          version: 1,
          last_modified: new Date(),
          hash: hash,
          sync_status: 'synced',
          created_at: new Date(),
          updated_at: new Date()
        });
        await version.save();
      }

      return version;
    } catch (error) {
      throw new Error(`Error obteniendo/creando versión: ${error.message}`);
    }
  }

  /**
   * Incrementa versión de un registro
   */
  async incrementVersion(collectionName, recordId, modifiedBy, deviceId, data) {
    try {
      const version = await this.getOrCreateVersion(collectionName, recordId, data);
      const newHash = this.calculateHash(data);
      
      // Si el hash cambió, incrementar versión
      if (version.hash !== newHash) {
        version.version += 1;
        version.last_modified = new Date();
        version.last_modified_by = modifiedBy;
        version.device_id = deviceId;
        version.hash = newHash;
        version.sync_status = 'synced';
        version.updated_at = new Date();
        await version.save();
      }

      return version;
    } catch (error) {
      throw new Error(`Error incrementando versión: ${error.message}`);
    }
  }

  /**
   * Registra o actualiza dispositivo
   */
  async registerDevice(deviceId, deviceInfo = {}) {
    try {
      const { v4: uuidv4 } = require('uuid');
      
      let device = await this.DeviceSync.findOne({ device_id: deviceId });
      
      if (!device) {
        device = new this.DeviceSync({
          _id: uuidv4(),
          device_id: deviceId,
          device_name: deviceInfo.device_name || 'Unknown',
          device_type: deviceInfo.device_type || 'mobile',
          app_version: deviceInfo.app_version || '1.0.0',
          last_sync: new Date(),
          last_sync_success: true,
          sync_token: this.generateSyncToken(),
          pending_changes: 0,
          conflict_count: 0,
          created_at: new Date(),
          updated_at: new Date()
        });
      } else {
        device.device_name = deviceInfo.device_name || device.device_name;
        device.device_type = deviceInfo.device_type || device.device_type;
        device.app_version = deviceInfo.app_version || device.app_version;
        device.updated_at = new Date();
      }

      await device.save();
      return device;
    } catch (error) {
      throw new Error(`Error registrando dispositivo: ${error.message}`);
    }
  }

  /**
   * Genera token de sincronización
   */
  generateSyncToken() {
    return crypto.randomBytes(32).toString('hex');
  }

  /**
   * Obtiene cambios desde el servidor (pull)
   */
  async getServerChanges(deviceId, lastSyncTimestamp, collections = []) {
    try {
      const device = await this.DeviceSync.findOne({ device_id: deviceId });
      if (!device) {
        throw new Error('Dispositivo no registrado');
      }

      const changes = [];
      const lastSync = lastSyncTimestamp ? new Date(lastSyncTimestamp) : new Date(0);

      // Obtener versiones modificadas después de lastSync
      const modifiedVersions = await this.DataVersion.find({
        last_modified: { $gt: lastSync },
        sync_status: { $ne: 'conflict' }
      }).sort({ last_modified: 1 });

      // Agrupar por colección y obtener datos
      const collectionsToSync = collections.length > 0 ? collections : ['asistencias', 'presencia'];
      
      for (const version of modifiedVersions) {
        if (collectionsToSync.includes(version.collection_name)) {
          // Obtener datos del registro
          const recordData = await this.getRecordData(version.collection_name, version.record_id);
          
          if (recordData) {
            changes.push({
              collection: version.collection_name,
              record_id: version.record_id,
              operation: 'update',
              data: recordData,
              version: version.version,
              last_modified: version.last_modified,
              hash: version.hash
            });
          }
        }
      }

      // Actualizar último sync del dispositivo
      device.last_sync = new Date();
      device.last_sync_success = true;
      await device.save();

      return {
        changes,
        sync_token: device.sync_token,
        server_timestamp: new Date()
      };
    } catch (error) {
      throw new Error(`Error obteniendo cambios del servidor: ${error.message}`);
    }
  }

  /**
   * Obtiene datos de un registro según la colección
   */
  async getRecordData(collectionName, recordId) {
    try {
      let Model;
      switch (collectionName) {
        case 'asistencias':
          Model = this.Asistencia;
          break;
        case 'presencia':
          Model = this.Presencia;
          break;
        default:
          return null;
      }

      const record = await Model.findById(recordId).lean();
      return record;
    } catch (error) {
      return null;
    }
  }

  /**
   * Sube cambios desde el cliente (push)
   */
  async uploadClientChanges(deviceId, changes) {
    try {
      const device = await this.DeviceSync.findOne({ device_id: deviceId });
      if (!device) {
        throw new Error('Dispositivo no registrado');
      }

      const results = {
        synced: [],
        conflicts: [],
        errors: []
      };

      for (const change of changes) {
        try {
          const result = await this.processChange(deviceId, change);
          
          if (result.status === 'conflict') {
            results.conflicts.push(result);
          } else if (result.status === 'error') {
            results.errors.push(result);
          } else {
            results.synced.push(result);
          }
        } catch (error) {
          results.errors.push({
            record_id: change.record_id,
            collection: change.collection,
            error: error.message
          });
        }
      }

      // Actualizar estadísticas del dispositivo
      device.pending_changes = results.conflicts.length;
      device.conflict_count = (device.conflict_count || 0) + results.conflicts.length;
      device.last_sync = new Date();
      device.last_sync_success = results.errors.length === 0;
      await device.save();

      return results;
    } catch (error) {
      throw new Error(`Error subiendo cambios del cliente: ${error.message}`);
    }
  }

  /**
   * Procesa un cambio individual
   */
  async processChange(deviceId, change) {
    try {
      const { collection, record_id, operation, data, version, hash } = change;

      // Obtener versión actual del servidor
      const serverVersion = await this.getOrCreateVersion(collection, record_id);

      // Verificar conflictos
      if (serverVersion.version > version) {
        // Conflicto: el servidor tiene una versión más reciente
        return await this.handleConflict(deviceId, collection, record_id, change, serverVersion);
      }

      // Aplicar cambio
      let result;
      switch (operation) {
        case 'create':
          result = await this.applyCreate(collection, data, deviceId);
          break;
        case 'update':
          result = await this.applyUpdate(collection, record_id, data, deviceId);
          break;
        case 'delete':
          result = await this.applyDelete(collection, record_id, deviceId);
          break;
        default:
          throw new Error(`Operación no soportada: ${operation}`);
      }

      // Actualizar versión
      if (result.success) {
        await this.incrementVersion(collection, record_id, data?.guardia_id || deviceId, deviceId, result.data);
      }

      return {
        status: 'synced',
        record_id: record_id,
        collection: collection,
        version: serverVersion.version + 1
      };
    } catch (error) {
      return {
        status: 'error',
        record_id: change.record_id,
        collection: change.collection,
        error: error.message
      };
    }
  }

  /**
   * Maneja conflictos
   */
  async handleConflict(deviceId, collection, recordId, clientChange, serverVersion) {
    try {
      // Obtener datos del servidor
      const serverData = await this.getRecordData(collection, recordId);

      // Marcar versión como conflicto
      serverVersion.sync_status = 'conflict';
      serverVersion.conflict_resolution = {
        strategy: 'last_write_wins', // Por defecto
        resolution_data: {
          client_data: clientChange.data,
          server_data: serverData,
          client_version: clientChange.version,
          server_version: serverVersion.version
        }
      };
      await serverVersion.save();

      // Guardar cambio pendiente para resolución
      const { v4: uuidv4 } = require('uuid');
      const pendingChange = new this.PendingChange({
        _id: uuidv4(),
        device_id: deviceId,
        collection_name: collection,
        record_id: recordId,
        operation: clientChange.operation,
        data: clientChange.data,
        version: clientChange.version,
        timestamp: new Date(),
        status: 'conflict',
        created_at: new Date()
      });
      await pendingChange.save();

      return {
        status: 'conflict',
        record_id: recordId,
        collection: collection,
        client_version: clientChange.version,
        server_version: serverVersion.version,
        server_data: serverData,
        conflict_id: pendingChange._id
      };
    } catch (error) {
      throw new Error(`Error manejando conflicto: ${error.message}`);
    }
  }

  /**
   * Aplica operación create
   */
  async applyCreate(collection, data, deviceId) {
    try {
      let Model;
      switch (collection) {
        case 'asistencias':
          Model = this.Asistencia;
          break;
        case 'presencia':
          Model = this.Presencia;
          break;
        default:
          throw new Error(`Colección no soportada: ${collection}`);
      }

      const record = new Model(data);
      await record.save();

      return {
        success: true,
        data: record.toObject()
      };
    } catch (error) {
      throw new Error(`Error aplicando create: ${error.message}`);
    }
  }

  /**
   * Aplica operación update
   */
  async applyUpdate(collection, recordId, data, deviceId) {
    try {
      let Model;
      switch (collection) {
        case 'asistencias':
          Model = this.Asistencia;
          break;
        case 'presencia':
          Model = this.Presencia;
          break;
        default:
          throw new Error(`Colección no soportada: ${collection}`);
      }

      const record = await Model.findByIdAndUpdate(recordId, data, { new: true });
      if (!record) {
        throw new Error('Registro no encontrado');
      }

      return {
        success: true,
        data: record.toObject()
      };
    } catch (error) {
      throw new Error(`Error aplicando update: ${error.message}`);
    }
  }

  /**
   * Aplica operación delete
   */
  async applyDelete(collection, recordId, deviceId) {
    try {
      let Model;
      switch (collection) {
        case 'asistencias':
          Model = this.Asistencia;
          break;
        case 'presencia':
          Model = this.Presencia;
          break;
        default:
          throw new Error(`Colección no soportada: ${collection}`);
      }

      const record = await Model.findByIdAndDelete(recordId);
      if (!record) {
        throw new Error('Registro no encontrado');
      }

      return {
        success: true,
        data: null
      };
    } catch (error) {
      throw new Error(`Error aplicando delete: ${error.message}`);
    }
  }

  /**
   * Resuelve un conflicto
   */
  async resolveConflict(conflictId, resolutionStrategy, resolvedBy, resolutionData = null) {
    try {
      const pendingChange = await this.PendingChange.findById(conflictId);
      if (!pendingChange || pendingChange.status !== 'conflict') {
        throw new Error('Conflicto no encontrado o ya resuelto');
      }

      const version = await this.DataVersion.findOne({
        collection_name: pendingChange.collection_name,
        record_id: pendingChange.record_id
      });

      if (!version) {
        throw new Error('Versión no encontrada');
      }

      let finalData;
      switch (resolutionStrategy) {
        case 'server_wins':
          // Mantener datos del servidor
          finalData = version.conflict_resolution.resolution_data.server_data;
          break;
        case 'client_wins':
          // Usar datos del cliente
          finalData = pendingChange.data;
          break;
        case 'merge':
          // Fusionar datos (requiere lógica específica)
          finalData = resolutionData || this.mergeData(
            version.conflict_resolution.resolution_data.server_data,
            pendingChange.data
          );
          break;
        case 'last_write_wins':
          // Usar el más reciente
          const clientTime = new Date(pendingChange.timestamp);
          const serverTime = version.last_modified;
          finalData = clientTime > serverTime ? pendingChange.data : version.conflict_resolution.resolution_data.server_data;
          break;
        default:
          throw new Error(`Estrategia de resolución no soportada: ${resolutionStrategy}`);
      }

      // Aplicar resolución
      await this.applyUpdate(
        pendingChange.collection_name,
        pendingChange.record_id,
        finalData,
        resolvedBy
      );

      // Actualizar versión
      version.sync_status = 'resolved';
      version.conflict_resolution.strategy = resolutionStrategy;
      version.conflict_resolution.resolved_by = resolvedBy;
      version.conflict_resolution.resolved_at = new Date();
      version.conflict_resolution.resolution_data = finalData;
      await version.save();

      // Marcar cambio como resuelto
      pendingChange.status = 'synced';
      await pendingChange.save();

      return {
        success: true,
        record_id: pendingChange.record_id,
        collection: pendingChange.collection_name,
        version: version.version + 1
      };
    } catch (error) {
      throw new Error(`Error resolviendo conflicto: ${error.message}`);
    }
  }

  /**
   * Fusiona datos de conflicto (merge básico)
   */
  mergeData(serverData, clientData) {
    // Merge simple: priorizar datos del cliente pero mantener campos del servidor que no existen en cliente
    return {
      ...serverData,
      ...clientData,
      // Mantener timestamps del servidor para campos de sistema
      fecha_creacion: serverData.fecha_creacion || clientData.fecha_creacion,
      fecha_actualizacion: new Date()
    };
  }

  /**
   * Obtiene conflictos pendientes
   */
  async getPendingConflicts(deviceId = null) {
    try {
      const query = { status: 'conflict' };
      if (deviceId) {
        query.device_id = deviceId;
      }

      const conflicts = await this.PendingChange.find(query)
        .sort({ timestamp: -1 })
        .lean();

      // Enriquecer con información de versión
      const enrichedConflicts = await Promise.all(conflicts.map(async (conflict) => {
        const version = await this.DataVersion.findOne({
          collection_name: conflict.collection_name,
          record_id: conflict.record_id
        });

        return {
          ...conflict,
          server_version: version?.version,
          server_data: version?.conflict_resolution?.resolution_data?.server_data
        };
      }));

      return enrichedConflicts;
    } catch (error) {
      throw new Error(`Error obteniendo conflictos: ${error.message}`);
    }
  }

  /**
   * Sincronización completa bidireccional
   */
  async performBidirectionalSync(deviceId, deviceInfo, lastSyncTimestamp, clientChanges) {
    try {
      // Registrar dispositivo
      await this.registerDevice(deviceId, deviceInfo);

      // Obtener cambios del servidor (pull)
      const serverChanges = await this.getServerChanges(deviceId, lastSyncTimestamp);

      // Subir cambios del cliente (push)
      const uploadResults = await this.uploadClientChanges(deviceId, clientChanges || []);

      return {
        server_changes: serverChanges,
        upload_results: uploadResults,
        sync_timestamp: new Date()
      };
    } catch (error) {
      throw new Error(`Error en sincronización bidireccional: ${error.message}`);
    }
  }
}

module.exports = BidirectionalSyncService;

